### DISCUSSION TITLE
Mapping through list of different text enries, how to add translations?

### DISCUSSION THREAD

**Discussion by BearCooder** [amannn/next-intl#535] [Q&A] [answered]
Hi! I hit a small issue with adding translations to every component with next-intl and I am not sure how to approach this.
I have a component where I map through my config where I store some text content. It looks more or less like this:
```
const contents: Content[] = [
  {
    id: 0,
    photo: photo-1,
    title: 'some text',
    description: 'some more text',
    introduction: 'some more text',
  },
   ..
   ..
   ..
 ];

export default contents;
```
The contents config is a separate file with around 50 entries. 
Right now I import the contents config into my component and map through the whole content and return the things to display it. 

But here comes my problem now. How would be the best possible way to apply translations with next-intl in this case?

Usually I would add my text into the messages/de.json etc and then `import { useTranslations } from 'next-intl';` then   `const t = useTranslations('Contact');` and add it like this `{t('header')}`.

But here I have so many entries that I map through and return, how can I add translations dynamically to every element that I map and return?

I was thinking to add the translations in the messages folder to the respective json files. Then import useTranslations into my contents config and add the translations directly here. But with the current structure of my contents config it is not possible.

**Comment by BearCooder** [comment]
Hey @amannn is there a way for "dynamic" translations?

I have in my example a description "with some text" and the description is for every object different. I map through and return then _description_. How could I assign the translations to description when the text string is different for every object in my list?

**Comment by xasanof17** [comment]
I just encountered such a situation and asked gpt there is [link](https://chat.openai.com/share/e745120c-480d-4027-8020-384230a9ee31) than i have used it. so i think we can't use it dynamic when i use gpt's answer, you can see [it](https://github.com/amannn/next-intl/discussions/554).
![code](https://github.com/amannn/next-intl/assets/93818267/364ea34a-40d5-4892-b6dd-5cbb71202608)
when i create featuresCard component and put it you can see the result:)
![image](https://github.com/amannn/next-intl/assets/93818267/36662277-0b4f-4da6-bc44-4bf24c0e3c9f)
how can we fix it i don't know

**Reply by xasanof17** [reply]
i think we have to use the same component and give object keys with prop)

**Comment by BearCooder** [comment]
Hello @amannn I created a minimal working Example on [Codesandbox](https://codesandbox.io/p/github/BearCooder/next-intl/main?import=true&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clnjhb29u00072a6afc0w9nxy%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clnjhb29u00032a6attj1j7q8%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clnjhb29u00052a6atrbjjft3%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clnjhb29u00062a6al6r28o4u%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B60%252C40%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clnjhb29u00032a6attj1j7q8%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clnjhb29u00022a6a2iua0ozh%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fapp%252F%255Blocale%255D%252Fglobals.css%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clnjhb29u00032a6attj1j7q8%2522%252C%2522activeTabId%2522%253A%2522clnjhb29u00022a6a2iua0ozh%2522%257D%252C%2522clnjhb29u00062a6al6r28o4u%2522%253A%257B%2522id%2522%253A%2522clnjhb29u00062a6al6r28o4u%2522%252C%2522activeTabId%2522%253A%2522clnjhbif0009g2a6aspfzxbnd%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clnjhbif0009g2a6aspfzxbnd%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%2522%257D%255D%257D%252C%2522clnjhb29u00052a6atrbjjft3%2522%253A%257B%2522id%2522%253A%2522clnjhb29u00052a6atrbjjft3%2522%252C%2522activeTabId%2522%253A%2522clnjhb29u00042a6adc1rq27m%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clnjhb29u00042a6adc1rq27m%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clnjhsah30019e5gfan13e49r%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522id%2522%253A%2522clnjhbfgz007x2a6af19upxte%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

On the main page I have two buttons, EN, DE to change the language. And then I can click on "go to products" button to see the products. And here I get the error: 
![grafik](https://github.com/amannn/next-intl/assets/104575503/17402b4f-0b7d-4b94-994b-86576db103bb)

I think I have a solution for all the different descriptions but it does not work, can you please take a look into ma dataconfig and the messages json?

**Reply by amannn** [reply, marked as answer]
There are multiple errors in your app:
1. You're calling a React component as a function in `Gallery.ts`. You need to use the file extension `.tsx` and use the JSX syntax to create elements.
2. You're wrapping the translated string `descriptionNature` in an unnecessary string template with curly brackets inside.
3. You don't return from your `.map` call in `products/page.tsx`.

Here's a basic fixed version of your CodeSandbox: https://codesandbox.io/p/github/BearCooder/next-intl/csb-n38mpp/draft/rough-worker?file=%2FdataConfig%2FGallery.tsx%3A19%2C38.

These are general React problems and don't have anything to do with `next-intl`. I'd kindly ask you to refer to the React and Next.js docs about general usage of the framework as I'm unfortunately not able to provide personal assistance with app development. I hope you can understand!

**Reply by amannn** [reply]
There's btw. also a React Discord channel where as far as I know people can ask for support: https://discord.com/invite/reactiflux

**Reply by BearCooder** [reply]
I am very sorry and I will join the discord. Thank you very much for you help and I will limit it to next-intl questions.
